service/matching/db_task_manager.go

```diff
@@ -0,0 +1,258 @@
+	"go.temporal.io/server/service/worker/scanner/taskqueue"
+)
+
+const (
+	dbTaskFlushInterval = 24 * time.Millisecond
+
+	dbTaskUpdateAckInterval   = time.Minute
+	dbTaskUpdateQueueInterval = time.Minute
+)
+
+type (
+	dbTaskManager struct {
+		status        int32
+		store         persistence.TaskManager
+		taskOwnership *dbTaskOwnershipImpl
+		taskFlusher   *dbTaskFlusher
+		taskTracker   *dbTaskTracker
+
+		dispatchTaskFn func(*internalTask) error
+		finishTaskFn   func(*persistencespb.AllocatedTaskInfo, error)
+		logger         log.Logger
+
+		shutdownChan          chan struct{}
+		writeNotificationChan chan struct{}
+		dispatchChan          chan struct{}
+	}
+)
+
+func newDBTaskManager(
+	taskQueueKey persistence.TaskQueueKey,
+	taskQueueKind enumspb.TaskQueueKind,
+	taskIDRangeSize int64,
+	store persistence.TaskManager,
+	logger log.Logger,
+	dispatchTaskFn func(*internalTask) error,
+	finishTaskFn func(*persistencespb.AllocatedTaskInfo, error),
+) (*dbTaskManager, error) {
+	taskOwnership := newDBTaskOwnership(
+		taskQueueKey,
+		taskQueueKind,
+		taskIDRangeSize,
+		store,
+		logger,
+	)
+	if err := taskOwnership.takeTaskQueueOwnership(); err != nil {
+		return nil, err
+	}
+
+	return &dbTaskManager{
+		status:        common.DaemonStatusInitialized,
+		store:         store,
+		taskOwnership: taskOwnership,
+		taskFlusher: newDBTaskFlusher(
+			taskQueueKey,
+			taskOwnership,
+			logger,
+		),
+		taskTracker: newDBTaskTracker(
+			taskQueueKey,
+			store,
+			taskOwnership.getAckedTaskID(),
+			logger,
+		),
+		dispatchTaskFn: dispatchTaskFn,
+		finishTaskFn:   finishTaskFn,
+		logger:         logger,
+
+		shutdownChan: make(chan struct{}),
+		dispatchChan: make(chan struct{}, 1),
+	}, nil
+}
+
+func (d *dbTaskManager) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&d.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	d.SignalDispatch()
+	go d.readerEventLoop()
+	go d.writerEventLoop()
+}
+
+func (d *dbTaskManager) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&d.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(d.shutdownChan)
+}
+
+func (d *dbTaskManager) SignalDispatch() {
+	select {
+	case d.dispatchChan <- struct{}{}:
+	default: // channel already has an event, don't block
+	}
+}
+
+func (d *dbTaskManager) isStopped() bool {
+	return atomic.LoadInt32(&d.status) == common.DaemonStatusStopped
+}
+
+func (d *dbTaskManager) writerEventLoop() {
+	updateQueueTicker := time.NewTicker(dbTaskUpdateQueueInterval)
+	defer updateQueueTicker.Stop()
+	// TODO we should impl a more efficient method to
+	//  buffer & wait for max duration
+	//  right now simply just flush every dbTaskFlushInterval
+	flushTicker := time.NewTicker(dbTaskFlushInterval)
+	defer flushTicker.Stop()
+
+	for {
+		if d.isStopped() {
+			return
+		}
+
+		select {
+		case <-d.shutdownChan:
+			return
+		case <-d.taskOwnership.getShutdownChan():
+			d.Stop()
```

Comment:
is it possible to use only one chan for shutdown?

Reply:
there is one go routine for read path, one for write path